### ISSUE TITLE
Add ignoreBeforeAtRules: [] to block-closing-brace-newline-after

### ISSUE THREAD

**Issue by srawlins** [stylelint/stylelint#4144]
> Clearly describe the bug

The [block-closing-brace-newline-after](https://stylelint.io/user-guide/rules/block-closing-brace-newline-after/#optional-secondary-options) rule has a neat [ignoreAtRules](https://stylelint.io/user-guide/rules/block-closing-brace-newline-after/#optional-secondary-options) option, which allows non-newline characters after an at-rule-associated `}`. The examples given are:

```scss
@if ($var) {
  color: pink;
} @else if ($var2) {
  color: red;
} @else {
  color: blue;
}
```

```scss
@if ($var) { color: pink; } @else { color: blue; }
```

This ignores too many violations. E.g. this also becomes considered clean code:

```scss
@if ($var) { color: pink; } p { margin: 2em; }
```

The `ignoreAtRules` option should only allow more at-rules after the `}` attached to an ignored at-rule. I can't conceive of why one would allow a declaration after such an `}`, or a rule, etc. I think the option was written specifically to support Sass's `@if`/`@else`.

> Which rule, if any, is the bug related to?

`block-closing-brace-newline-after`

> What CSS is needed to reproduce the bug?

```scss
@if ($var) { color: pink; } p { margin: 2em; }
```

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "block-closing-brace-newline-after": ["always", "ignoreAtRules": ["if", "else"]]
  }
}
```

> Which version of stylelint are you using?

`9.6.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

Node.js API using `cli.js`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to Sass nested properties.

> What did you expect to happen?

block-closing-brace-newline-after to be flagged for the rule after the `}` associated with the `@if`.

> What actually happened (e.g. what warnings or errors did you get)?

No warnings were flagged.

If this is accepted as a bug, I can likely write a PR to fix it.

**Comment by evilebottnawi** [comment]
For specify sass/scss rules better open issue in stylelint-scss, think option ignore any newlines for `at-rules` in your list

**Comment by srawlins** [comment]
Hi @evilebottnawi, I'm not sure what you are suggesting. block-closing-brace-newline-after is a stylelint rule, not a stylelint-scss rule. What is an example use case for the `ignoreAtRules` option that pertains to CSS?

I cannot think of a reason that a user would want the `ignoreAtRules` option to allow something other than an at-rule to follow a `}`.

**Comment by evilebottnawi** [comment]
@srawlins sometimes you just need ignore something

Good practice using `ignoreAtRules` for ignore non standard at-rules and implement rule in `stylelint-scss` for `if/else` style

**Comment by srawlins** [comment]
@davidtheclark or @jeddy3, would you mind weighing in here on the purpose of the `ignoreAtRules` option? I see in https://github.com/stylelint/stylelint/issues/561 a lot of good discussion and back-and-forth on what this is designed for. It only ever seems to be to allow newlines before `{`, and more at-rules after `}`.

@evilebottnawi I still am not sure what you are suggesting. I would like a rule which reports when a non-newline follows an ar-rule's `}`, such as:

```scss
p.a { margin: 2em; } p.b { margin: 2em; }
// report here ------------^

@if ($var) {
  color: pink;
} p { margin: 2em; }
// ^-- report here
```

The stylelint tool has a great rule for that: `block-closing-brace-newline-after`. Unfortunately, there is an exception I would like to make to the rule: other at-rules may follow a `}`, such as:

```scss
@if ($var) {
  color: pink;
} @else if ($var2) {
  color: red;
} @else {
  color: blue;
}
```

The stylelint tool has a great option for that exception: `ignoreAtRules`. However, `ignoreAtRules` allows any non-newline character (accidentally, I'm sure) after `}`, not just other at-rules. `ignoreAtRules` was written to allow other at-rules to be specified immediately after a `}`.

This is how I am interpreting your suggestion, @evilebottnawi, but I might be wrong: `ignoreAtRules` is not working for my use case; you posit that there is nothing to fix, since I have a Sass problem, not a CSS problem. You are suggesting that I fix this problem with stylelint-scss. You are suggesting that I copy and paste all of the `block-closing-brace-newline-after` code from stylelint, and paste it into stylelint-scss, make the desired change to `ignoreAtRules`, then disable the `block-closing-brace-newline-after` rule and enable my new `scss/block-closing-brace-newline-after` rule. Is this correct?

The fix I am proposing for `block-closing-brace-newline-after` is something like:

```diff
- if (optionsMatches(options, "ignoreAtRules", statement.name)) {
-   return;
- }
- 
 const nextNode = statement.next();

 if (!nextNode) {
   return;
 }
+ 
+ if (optionsMatches(options, "ignoreAtRules", statement.name) &&
+     optionsMatches(options, "ignoreAtRules", nextNode.name)) {
+   return;
+ }
```

**Comment by evilebottnawi** [comment]
stylelint doesn't report on any non standard(non css) stuff, so if you want get warnings/errors for sass/scss stuff you need create rule/open issue in `stylelint-scss`

**Comment by nex3** [comment]
Hey @evilebottnawi! We could just update this within Google, but I'd really like to find a way to make this work for all Sass users. I hear you on not wanting to add Sass-specific logic to the general stylelint package, but this may even become relevant to all *CSS* users—there's discussion of adding something like `@else` in CSS itself (https://github.com/w3c/csswg-drafts/issues/112).

What if we added a new configuration field, `ignoreBeforeAtRules`, that takes a list of at-rule names (just like `ignoreAtRules`) but only ignores whitespace after a brace *and in front of those at-rules*? For example, `"ignoreBeforeAtRules": ["else"]` would allow


```scss
@if ($var) {
  color: pink;
} @else if ($var2) {
  color: red;
} @else {
  color: blue;
}
```

and

```scss
@media (width < 28em) {
  // ...
} @else {
  // ...
}
```

but would complain about

```scss
@if ($var) { color: pink; } p { margin: 2em; }
```

**Comment by hudochenkov** [comment]
I believe non of our rules looks for nodes (properties, rules, etc.) _after_ current node that is being checked. All rules looks at current node, and sometime at previous node.

Because of this add suggested `ignoreBeforeAtRules` won't be possible as it will require to look to the following. Well technically it would be possible, but it won't be done in stylelint core.

> this may even become relevant to all CSS users

When it happen we can reevaluate this, until then this feature targeted for preprocessors (and very niche even for preprocessor users). I more than sure users start their CSS rules from the new line, and very-very rarely CSS rule could appear on the same line as the end of `@if/@else`.

> You are suggesting that I copy and paste all of the block-closing-brace-newline-after code from stylelint, and paste it into stylelint-scss, make the desired change to ignoreAtRules, then disable the block-closing-brace-newline-after rule and enable my new scss/block-closing-brace-newline-after rule. Is this correct?

I would say this is the way to go.

**Comment by nex3** [comment]
> I believe non of our rules looks for nodes (properties, rules, etc.) _after_ current node that is being checked. All rules looks at current node, and sometime at previous node.
> 
> Because of this add suggested `ignoreBeforeAtRules` won't be possible as it will require to look to the following. Well technically it would be possible, but it won't be done in stylelint core.

Why is this a reason not to add it? PostCSS makes this information available, so why not use it?

**Comment by hudochenkov** [comment]
I can't say for sure. I was here from the beginning of the project. Probably to avoid rule conflicts. Like with CSS: it's easier to compose block when they have `margin-bottom` only, rather when some has `margin-top`, some `margin-bottom`.

**Comment by nex3** [comment]
If there's not a concrete use-case that this breaks, can you add an exception here? It would be very useful for many users of this package.

**Comment by hudochenkov** [comment]
I wouldn't say adding this is a bad idea, but the value is almost non-existent. I would expect that only few people once in many month do something like:

```scss
@if ($var) { color: pink; } p { margin: 2em; }
```

**Comment by jeddy3** [comment]
>  @jeddy3, would you mind weighing in here on the purpose of the ignoreAtRules option?

Yep, no problem. I'm just back from my hols and can chime in now.

@nex3 and @srawlins, am I right in thinking that a `ignoreBeforeAtRules` option will nicely complement the [`@if`](https://github.com/kristerkari/stylelint-scss#-if) and [`@else`](https://github.com/kristerkari/stylelint-scss#-else) rules in [stylelint-scss](https://github.com/kristerkari/stylelint-scss)? And it will allow a user to apply those rules to enforce a strict style convention for their `@if` and `@else` blocks _while also_ protecting them against the following?

```scss
@if ($var) { color: pink; } p { margin: 2em; }
```

If so, then I think we should add the option. 

When we first started on stylelint we set out to [support an abundance of code styles](https://stylelint.io/user-guide/about-rules#newline-space-before-and-newline-space-after-rules) by creating all the `*-space-*` and `*-newline-*` rules with their `*-single-line` and `*-multi-line` options. We did this so that users could enforce a strict set of style conventions and _stop any anomalies (like the code example above) slipping through_. I think this option aligns to that vision (while also assisting the stylelint-scss plugin maintainers).

(I think we'll want to add a new `ignoreBeforeAtRules` option, rather than change the behaviour of the existing `ignoreAtRules` one. So that the `ignoreAtRules` option is consistent with those in other rules and in case anyone is relying on its current behaviour.)

>>>  Because of this add suggested ignoreBeforeAtRules won't be possible as it will require to look to the following. Well technically it would be possible, but it won't be done in stylelint core.

>>  Why is this a reason not to add it?

> I can't say for sure. Probably to avoid rule conflicts

I don't believe there's a reason why we haven't created an option that inspects the next node yet. It might be because we simply haven't needed to until now. Let's just make sure we add a test where there isn't a next node so that we guard against it.

---

Unless there are any objections, I'll label this up as a new option and open it out to a PR.

**Comment by nex3** [comment]
@jeddy3 Sounds good to me!

**Comment by jeddy3** [comment]
I've now labelled the issue as "help wanted". @nex3 and @srawlins Feel free to contribute a PR, thanks.

**Comment by nex3** [comment]
@rambleraptor can you look into this?

**Comment by rambleraptor** [comment]
yeah, I'm on it!

**Comment by ermik** [comment]
Is there an example of how to implement this sort of thing?

**Comment by jeddy3** [comment]
@ermik Thanks for wanting to [contribute](https://stylelint.io/contributing)

There are [steps on how to add a new option](https://stylelint.io/developer-guide/rules#add-an-option-to-a-rule) in the Developer guide.

[This is an example of a recent pull request](https://github.com/stylelint/stylelint/pull/5520/files) to add an option to a rule. 

Please branch off `v14` rather than `master`.

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and add this option.
